### PR TITLE
Change template arguments in DerivativeApproximation::approximate_derivative_tensor

### LINKED ISSUE #3979: DerivativeApproximation::approximate_derivative_tensor requires template arguments
With relation to [this forum post](https://groups.google.com/forum/#!topic/dealii/2KEDTb1jDjo), the following MWE does not compile without additional template arguments being sent to `DerivativeApproximation::approximate_derivative_tensor`
```c++
#include <deal.II/base/tensor.h>
#include <deal.II/dofs/dof_handler.h>
#include <deal.II/dofs/dof_accessor.h>
#include <deal.II/grid/tria.h>
#include <deal.II/grid/tria_accessor.h>
#include <deal.II/lac/vector.h>
#include <deal.II/numerics/derivative_approximation.h>

using namespace dealii;

int main ()
{
  Triangulation<2,2> triangulation;
  DoFHandler<2,2> dof_handler;
  Vector<double> solution;

  Tensor<1,2> cell_grad;
  Tensor<2,2> cell_hessian;

  DoFHandler<2>::active_cell_iterator
  cell = dof_handler.begin_active(),
  endc = dof_handler.end();
  for (; cell!=endc; ++cell)
  {
    // Doesn't work
    DerivativeApproximation::approximate_derivative_tensor(dof_handler, solution, cell, cell_grad);
    DerivativeApproximation::approximate_derivative_tensor(dof_handler, solution, cell, cell_hessian);
    // Works
    // DerivativeApproximation::approximate_derivative_tensor<DoFHandler<2>,2,2>(dof_handler, solution, cell, cell_grad);
    // DerivativeApproximation::approximate_derivative_tensor<DoFHandler<2>,2,2>(dof_handler, solution, cell, cell_hessian);
  }

  return 0;
}
```

It is necessary to fully qualify the DoFHandler type, dim and spacedim because
```sh
<path to my test code>/derivative_approximation.cc:26:5: error: no matching function for call to 'approximate_derivative_tensor'
    DerivativeApproximation::approximate_derivative_tensor(dof_handler, solution, cell, cell_grad);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<path>/include/deal.II/numerics/derivative_approximation.h:260:3: note: candidate template ignored: couldn't infer template argument 'spacedim'
  approximate_derivative_tensor
  ^
<path>/include/deal.II/numerics/derivative_approximation.h:243:3: note: candidate function template not viable: requires at least 5 arguments, but 4 were provided
  approximate_derivative_tensor
  ^
```

This doesn't seem very user friendly. One possible fix is to change the template arguments from
```
template <typename DoFHandlerType, int dim, int spacedim, class InputVector, int order>
void approximate_derivative_tensor
```
to
```
template <template<int,int> class DoFHandlerType, int dim, int spacedim, class InputVector, int order>
void approximate_derivative_tensor
```
and the function argument from `const DoFHandlerType &dof` 
to `const DoFHandlerType<dim,spacedim> &dof` (I see that this is already done for `approximate_gradient`).

### PR DESCRIPTION
**Pull request by jppelteret**
Fixes #3979

### CODE REVIEW

**Comment by jppelteret** [comment]
What's a little bit strange is that the [test for this function](https://github.com/dealii/dealii/blob/master/tests/numerics/derivative_approximation_2.cc#L174-L199) didn't specify any template arguments, but they were definitely necessary in my MWE shown in #3979.

**Comment by bangerth** [comment]
It worked before because the test only tested the version with a mapping argument. The function you're changing doesn't have the mapping argument, and consequently there is no way to deduce that one template argument.

I guess this also means that the function in question may not be tested. Could you just clone the _2 test into a _3 (or similar) and remove the mapping argument in the call?

**Comment by jppelteret** [comment]
@bangerth Thanks for making that obvious observation on my behalf :-) I've added the new test that checks without the mapping.

**Comment by bangerth** [comment]
It took me a couple of minutes as well to see the problem. In fact I had already started typing a response saying that it is really bizarre that the compiler would accept the code!

File: doc/news/changes/incompatibilities/20170216Pelteret [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Changed: The template arguments to the 
+DerivativeApproximation::approximate_derivative_tensor() function have
+been modified. It is no longer necessary to define the <tt>dim<\tt>
```
Comment (bangerth) [inline doc/news/changes/incompatibilities/20170216Pelteret]: define -> explicitly specify

Comment (jppelteret) [inline doc/news/changes/incompatibilities/20170216Pelteret]: Done!

**Pull request merged**